CMakeLists.txt

```diff
@@ -224,16 +224,25 @@ endif()
 
 add_subdirectory(portable)
 
-add_library(freertos_kernel STATIC
+list(APPEND FREERTOS_KERNEL_SOURCES
     event_groups.c
     list.c
     queue.c
     stream_buffer.c
     tasks.c
     timers.c
+)
 
-    # If FREERTOS_HEAP is digit between 1 .. 5 - it is heap number, otherwise - it is path to custom heap source file
-    $<IF:$<BOOL:$<FILTER:${FREERTOS_HEAP},EXCLUDE,^[1-5]$>>,${FREERTOS_HEAP},portable/MemMang/heap_${FREERTOS_HEAP}.c>
+# Check if user requested to not inlude a heap implementation
+if(NOT DEFINED FREERTOS_DO_NOT_INCLUDE_HEAP)
```

Comment:
Perhaps we should just not include a heap implementation if FREERTOS_HEAP is not defined?

Reply:
I'm not sure it will compile without the heap definitions.  You will have linker issues with missing symbols.

If you do not want a heap then create a custom heap source that errors out whenever the heap is used/consumed by the API it defines, and do:

```cmake
set(FREERTOS_HEAP "heap_disabled.c")
```

Where `heap_disabled.c` is defined as:

```c

void * pvPortMalloc( size_t xWantedSize )
{
    (void)xWantedSize;

    #if ( configUSE_MALLOC_FAILED_HOOK == 1 )
    {
        vApplicationMallocFailedHook();
    }
    #endif

    return NULL;
}
/*-----------------------------------------------------------*/

void vPortFree( void * pv )
{
    ( void ) pv;
    /* Force an assert as it is invalid to call this function. */
    configASSERT( pv == NULL );
}
/*-----------------------------------------------------------*/

void vPortInitialiseBlocks( void )
{
}
/*-----------------------------------------------------------*/

size_t xPortGetFreeHeapSize( void )
{
    return 0;
}
```

If you want it in the FreeRTOS repo, then suggest making this an alternative heap implementation - as @paulbartell suggested and change the name to heap_0.c (NullOpt - a no-heap implementation of the heap API). - and place it with the other heap locations.  Then you'd also have to change:

```cmake
$<IF:$<BOOL:$<FILTER:${FREERTOS_HEAP},EXCLUDE,^[0-5]$>>,${FREERTOS_HEAP},portable/MemMang/heap_${FREERTOS_HEAP}.c>
```
To allow heap_0.c to be allowed.

Reply:
@paulbartell @jasonpcarroll  thoughts?

Reply:
Just a note since I opened the issue #594: It will compile without linker issues if `configSUPPORT_DYNAMIC_ALLOCATION` is set to 0 in the config header.

If adding the `heap_0.c` as suggested above, then I guess I would be relying on the linker to discard those sections. That is ok, but it would be neat in my opinion if that was the default. It seems weird to have to pick a heap implementation if the project is set up to not support dynamic allocation.

Reply:
I still stand by my previous comment - adding an additional FREERTOS_HEAP option for the enumeration is better than adding yet another configuration for the same feature.

Since there is a linkage between FREERTOS_HEAP and SUPPORT_DYNAMIC_ALLOCATION then I suggest you make the linkage within the config via [CMakeDependentOption](https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html) - so that it is explicitly known that if SUPPORT_DYNAMIC_ALLOCATION is set to 1, you must configure FREERTOS_HEAP.  Another method of fixing is to remove the SUPPORT_DYNAMIC_ALLOCATION and only use FREERTOS_HEAP=0 instead.

That way you have the error/issue identified even before compile - when building up the configuration instead.

Reply:
configSUPPORT_DYNAMIC_ALLOCATION should be sufficient to deciding to include a heap implementation.  The portable.h header should be changed so the heap functions are not declared when configSUPPORT_DYNAMIC_ALLOCATION==0.  It does make sense to add a dependent section to the CMakeLists to make a heap selection IF configSUPPORT_DYNAMIC_ALLOCATION==1 THOUGH, I would prefer that application decisions like heap & port be made at the application level CMake and not at the kernel level CMake.  This also has the advantage of a simpler path to a custom heap should one be desired.

Reply:
In the CMake you do not have access to configSUPPORT_DYNAMIC_ALLOCATION and configSUPPORT_STATIC_ALLOCATION (from FreeRTOSConfig.h), right? I made a PR to change the default behaviour of FREERTOS_HEAP , see https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/807 . (I was unaware that there were active pull requests attempting to address the same issue.).